Handle GWCS for specutils 2.0 compatibility

rosteen opened the pull request:
Adding better handling of GWCS in the Spectrum1D handler. Opening as draft since this may need a couple other changes/additions, and because it shouldn't be merged until specutils 2.0 is ready for release. I'll also add some test coverage for the new code.

pllim commented on setup.cfg:
```diff
@@ -20,7 +20,7 @@ install_requires =
     glue-core>=1.6.1
     regions>=0.4
-    specutils>=0.7
+    specutils>=2.0
```
This pin causes this branch refusing to install because specutils is still at `2.0.dev` that is technically less than 2.0 stable.

rosteen replied:
I think I'll do a 1.x release that adds the `Spectrum` class as an empty wrapper around `Spectrum1D`, so that people can update the name and now worry about errors in 1.x vs 2.0.

dhomeier replied:
Sounds good. Deprecation warnings should be ok, that's something we can catch in CI. But otherwise we'd probably have to abandon Python < 3.11 sooner than desired.

dhomeier replied:
That will still mean giving py38 and py39 support here, but I hope that's not too controversial.

pllim commented on setup.cfg:
```diff
@@ -20,7 +20,7 @@ install_requires =
     glue-core>=1.6.1
     regions>=0.4
-    specutils>=0.7
+    specutils>=2.0
```
This would let the install happen but would also cause this PR to not able to be merged because PyPI disallows dev pins for releases.

```suggestion
    specutils>=2.0.dev
```

dhomeier commented:
Should probably move the `-dev` jobs up from py310-py312 to py311-py313, and might have to include specreduce – is specreduce dev compatible with specutils 2.0?

dhomeier commented on .github/workflows/main.yml:
```diff
@@ -33,18 +33,16 @@ jobs:
         # Standard tests
-        - linux: py38-test
-        - linux: py39-test
+        - linux: py311-test
```
```suggestion
        - linux: py310-test
```
should keep at least one 3.10 job as we no longer can run it for dev

rosteen replied:
Whoops, good call. Thought I kept one.

dhomeier reviewed (commented):
Probably need to wait for astropy/specutils#1229 to be merged and a 1.20 released before the non-dev jobs can be properly tested.
I think the GWCS warnings can just be ignored in `filterwarnings` for the time being.
But `Spectrum` can apparently not simply be called with default args as a drop-in replacement for `Spectrum1D` – is that intended?

dhomeier commented:
I've tested this locally with specutils main after 1229 was merged, getting the
```python
>           data = Data(coords=PaddedSpectrumWCS(obj.wcs, obj.flux.ndim, obj.spectral_axis_index))
E           AttributeError: 'Spectrum' object has no attribute 'spectral_axis_index'
```
failures there as well.

dhomeier commented on setup.cfg:
```diff
@@ -20,7 +20,7 @@ install_requires =
     glue-core>=1.6.1
     regions>=0.4
-    specutils>=0.7
+    specutils>=2.0.0rc0
```
```suggestion
    specutils>=1.20
```
should set this so the py310 tests can continue to be run, and leave 2.0 for the -dev tests for now.

dhomeier commented:
So specreduce 1.20 does work here; dev may still be pulling in the 1.x branch though, but is failing on the gwcs deprecation warnings anyway.

astrofrog commented:
@rosteen could you rebase to pick up the latest CI changes in main?

Also to be clear, should this be merged before or after specutils 2.0 is merged?

dhomeier commented on glue_astronomy/translators/spectrum1d.py:
```diff
@@ -34,15 +35,16 @@
     # NOTE: This class could be updated to use CompoundLowLevelWCS from NDCube.
 
-    def __init__(self, wcs, ndim):
+    def __init__(self, wcs, ndim, spectral_axis_index=None):
```
Is there a reason not to support `move_spectral_axis` here as well? I suppose its functionality can be fully replaced by the right use of `spectral_axis_index`(?), but if people get used to one or the other...

rosteen replied:
Nope, it definitely should support that as well, just an oversight/order of when I was implementing things here and in specutils. I'll add it.

rosteen replied:
Actually, now that I look at it, I'm not sure why `spectral_axis_index` is here, I think I had started implementing something differently and then didn't end up using it. I think that neither needs to be here in `PaddedSpectrumWCS` but I'll double check. If anything the place it might make sense to add the option to use `move_spectral_axis` is in `to_object`.

dhomeier commented:
In the current state the specutils 1.2x transition version would be used, but dev tests still would not run on the 2.0 branch until that is merged – they are passing with both versions anyway.
I do think there should also be tests for passing the new `spectral_axis_index` and `move_spectral_axis` (if the latter can be used) through the translator, though those would have to discriminate between 1.x and 2.x.

dhomeier commented on glue_astronomy/translators/nddata.py:
```diff
@@ -190,6 +190,8 @@ def to_object(self, data_or_subset, attribute=None):
         else:
             data = data_or_subset
 
+        if attribute is None and "flux" in data.main_components:
+            attribute = "flux"
```
```suggestion
        if attribute is None and "flux" in data.main_components:
            attribute = "flux"
```
```suggestion
        if attribute is None and len(data.main_components) > 1 and "flux" in data.main_components:
            attribute = "flux"
```
Otherwise this would change behaviour in the one case of a single component that is named differently (though that's obviously not tested, so may be not really relevant). It might simplify things to handle this all inside `_get_attribute` at this point, though it should get some documentation here, e.g.
```python
        attribute : `glue.core.component_id.ComponentID`
            The attribute to use for the StdDevUncertainty data. If not specified,
            will use a single component or try to identify one named "flux".
```

dhomeier commented on glue_astronomy/translators/spectrum1d.py:
```diff
@@ -126,17 +127,47 @@ def serialized_classes(self):
+            # In this case we had spectral axis in pixels
+            spec_coord = test_world[data_ndim - 1 - spectral_axis_index]
+
+        return np.isclose(spec_coord[0], spec_coord[1]) and np.isclose(spec_coord[2], spec_coord[3])  # noqa
```
Will the default `rtol=1e-5` be enough here? Might actually already be of order ~1 resolution element for hires data.

dhomeier reviewed (commented):
Thanks, looks like things are taking shape now! I think the parts flagged by codecov give a good pointer where to add tests.

rosteen commented:
@dhomeier most of the areas that Codecov is currently flagging as uncovered are actually covered by tests when they run with specutils 2.0. I added a couple more checks for things that were really missing coverage, but I went ahead and marked this ready for review since I addressed your other comments.

dhomeier reviewed (approved):
Great, everything looks ready then – the way the legacy API is handled in the tests makes sense to me.
>  It looks like the last dev CI failure was due to regions, not specutils 🤔

Yes, though I could not reproduce the build failure on my (Intel) macOS. Strangely the header raising the error `npy_1_7_deprecated_api.h` does not even exist any longer on numpy 2.3. It seems regions is first installing its own numpy version in the build env, though the `numpy>=2.0.0rc1` dependency first pulls in either 2.3.0rc1 or 2.4.0.dev0 from nightly it then here ends up with 2.2.6...
Anyway unrelated; may have to move the macos-dev job to 3.12, which has regions wheels, until the 0.11 release is out.

dhomeier commented on tox.ini:
```diff
@@ -14,7 +14,8 @@ deps =
     dev: scipy>=0.0.dev0
     dev: astropy>=0.0.dev0
-    dev: git+https://github.com/astropy/specutils.git
+    dev: specutils>=0.0dev0
```
```suggestion
    dev: git+https://github.com/astropy/regions.git
    dev: git+https://github.com/astropy/specutils.git
```
Revert to see if it works against the 2.0 branch now, and give regions dev a try for the macOS build.
